`ReplService.evalCode` does not contribute to TypeScript's view of the scope

### Expected Behavior

Executing code with `evalCode` to be equivalent to typing some code into the REPL's stdin

### Actual Behavior

The REPL believes any declared variables are simultaneously not declared and already declared

### Steps to reproduce the problem

```js
const tsNode = require("ts-node");

let repl = tsNode.createRepl();
let service = tsNode.create({...repl.evalAwarePartialHost});
repl.setService(service);
repl.start();

repl.evalCode(`let foo = "bar"; console.log(foo);`);
```

Observe the code has been executed as `bar` is printed to stdout.

Enter `foo` into stdin, incorrectly giving
`<repl>.ts:5:1 - error TS2304: Cannot find name 'foo'.`
(should print `bar`)

Enter `let foo = "baz";` into stdin giving the correct error, but with the wrong source text
```
<repl>.ts:1
"use strict"; void 0;
^

Uncaught SyntaxError: Identifier 'foo' has already been declared
```

### Specifications

```
ts-node v10.7.0
node v18.1.0
compiler v4.6.4
```

* Operating system and version:
Arch Linux

--- a/src/repl.ts
+++ b/src/repl.ts
@@ -207,6 +207,7 @@ export function createRepl(options: CreateReplOptions = {}) {
       state,
       input: code,
       context,
+      overrideIsCompletion: false,
     });
     assert(result.containsTopLevelAwait === false);
     return result.value;
@@ -512,6 +513,12 @@ function appendCompileAndEvalInput(options: {
   /** Enable top-level await but only if the TSNode service allows it. */
   enableTopLevelAwait?: boolean;
   context: Context | undefined;
+  /**
+   * Added so that `evalCode` can be guaranteed *not* to trigger the `isCompletion`
+   * codepath.  However, the `isCompletion` logic is ancient and maybe should be removed entirely.
+   * Nobody's looked at it in a long time.
+   */
+  overrideIsCompletion?: boolean;
 }): AppendCompileAndEvalInputResult {
   const {
     service,
@@ -519,6 +526,7 @@ function appendCompileAndEvalInput(options: {
     wrappedErr,
     enableTopLevelAwait = false,
     context,
+    overrideIsCompletion,
   } = options;
   let { input } = options;
 
@@ -533,7 +541,7 @@ function appendCompileAndEvalInput(options: {
   }
 
   const lines = state.lines;
-  const isCompletion = !/\n$/.test(input);
+  const isCompletion = overrideIsCompletion ?? !/\n$/.test(input);
   const undo = appendToEvalState(state, input);
   let output: string;
 

